src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,329 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/none.hpp>
+#include <boost/optional.hpp>
+#include <boost/optional/optional_io.hpp>
+#include <cstddef>
+
+#include "DataStructures/DenseMatrix.hpp"
+#include "DataStructures/DenseVector.hpp"
+#include "Informer/Verbosity.hpp"
+#include "NumericalAlgorithms/Convergence/Criteria.hpp"
+#include "NumericalAlgorithms/Convergence/HasConverged.hpp"
+#include "NumericalAlgorithms/Convergence/Reason.hpp"
+#include "NumericalAlgorithms/LinearSolver/InnerProduct.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace LinearSolver {
+namespace gmres_detail {
+
+// Perform an Arnoldi orthogonalization to find a new `operand` that is
+// orthogonal to all vectors in `basis_history`. Appends a new column to the
+// `orthogonalization_history` that holds the inner product of the intermediate
+// `operand` with each vector in the `basis_history` and itself.
+template <typename VarsType>
+void arnoldi_orthogonalize(
+    const gsl::not_null<VarsType*> operand,
+    const gsl::not_null<DenseMatrix<double>*> orthogonalization_history,
+    const std::vector<VarsType>& basis_history,
+    const size_t iteration) noexcept {
+  // Resize matrix and make sure the new entries that are not being filled below
+  // are zero.
+  orthogonalization_history->resize(iteration + 2, iteration + 1);
+  for (size_t j = 0; j < iteration; ++j) {
+    (*orthogonalization_history)(iteration + 1, j) = 0.;
+  }
+  // Arnoldi orthogonalization
+  for (size_t j = 0; j < iteration + 1; ++j) {
+    const double orthogonalization = inner_product(basis_history[j], *operand);
+    (*orthogonalization_history)(j, iteration) = orthogonalization;
+    *operand -= orthogonalization * basis_history[j];
+  }
+  (*orthogonalization_history)(iteration + 1, iteration) =
+      sqrt(inner_product(*operand, *operand));
+  *operand /= (*orthogonalization_history)(iteration + 1, iteration);
```

Comment:
This line can divide by zero.  I see it happen in the "Exact inverse preconditioner" test.

Reply:
@wthrowe pushed a fixup to handle this case, please take a look